contracts/contracts/TACoApplication.sol

```diff
@@ -0,0 +1,708 @@
+
+    /**
+    * @notice Signals that authorization was decreased involuntary
+    * @param stakingProvider Staking provider address
+    * @param fromAmount Previous amount of authorized tokens
+    * @param toAmount Amount of authorized tokens to decrease
+    */
+    event AuthorizationInvoluntaryDecreased(address indexed stakingProvider, uint96 fromAmount, uint96 toAmount);
+
+    /**
+    * @notice Signals that authorization decrease was requested for the staking provider
+    * @param stakingProvider Staking provider address
+    * @param fromAmount Current amount of authorized tokens
+    * @param toAmount Amount of authorization to decrease
+    */
+    event AuthorizationDecreaseRequested(address indexed stakingProvider, uint96 fromAmount, uint96 toAmount);
+
+    /**
+    * @notice Signals that authorization decrease was approved for the staking provider
+    * @param stakingProvider Staking provider address
+    * @param fromAmount Previous amount of authorized tokens
+    * @param toAmount Decreased amount of authorized tokens
+    */
+    event AuthorizationDecreaseApproved(address indexed stakingProvider, uint96 fromAmount, uint96 toAmount);
+
+    /**
+    * @notice Signals that authorization was resynchronized
+    * @param stakingProvider Staking provider address
+    * @param fromAmount Previous amount of authorized tokens
+    * @param toAmount Resynchronized amount of authorized tokens
+    */
+    event AuthorizationReSynchronized(address indexed stakingProvider, uint96 fromAmount, uint96 toAmount);
+
+    /**
+    * @notice Signals that the staking provider was slashed
+    * @param stakingProvider Staking provider address
+    * @param penalty Slashing penalty
+    * @param investigator Investigator address
+    * @param reward Value of reward provided to investigator (in units of T)
+    */
+    event Slashed(address indexed stakingProvider, uint256 penalty, address indexed investigator, uint256 reward);
+
+    /**
+    * @notice Signals that an operator was bonded to the staking provider
+    * @param stakingProvider Staking provider address
+    * @param operator Operator address
+    * @param previousOperator Previous operator address
+    * @param startTimestamp Timestamp bonding occurred
+    */
+    event OperatorBonded(
+        address indexed stakingProvider,
+        address indexed operator,
+        address indexed previousOperator,
+        uint256 startTimestamp
+    );
+
+    /**
+    * @notice Signals that an operator address is confirmed
+    * @param stakingProvider Staking provider address
+    * @param operator Operator address
+    */
+    event OperatorConfirmed(address indexed stakingProvider, address indexed operator);
+
+    struct StakingProviderInfo {
+        address operator;
+        bool operatorConfirmed;
+        uint64 operatorStartTimestamp;
+
+        uint96 authorized;
+        uint96 deauthorizing; // TODO real usage only in getActiveStakingProviders, maybe remove?
```

Comment:
I don't think it worth it to decide rn, because this relates to how we will slash, reward (automatic one) and other nuances which would be easier to obtain from using this contract